src/merklized_state/data_source.rs

```diff
@@ -0,0 +1,152 @@
+// see <https://www.gnu.org/licenses/>.
+
+//! Data for the [`merklized_state`](super) API.
+//!
+//! This module facilitates storing the state of a Merkle Tree at a specific point in time
+//! and provides methods for querying and reconstructing the snapshot.
+//!
+use ark_serialize::CanonicalDeserialize;
+use derive_more::{Display, From};
+
+use async_trait::async_trait;
+use commit::Commitment;
+use derivative::Derivative;
+use hotshot::traits::ValidatedState;
+use hotshot_types::{data::Leaf, traits::node_implementation::NodeType};
+use jf_primitives::{
+    circuit::merkle_tree::MembershipProof,
+    merkle_tree::{
+        prelude::{MerkleNode, MerklePath},
+        Element, Index, MerkleTreeScheme, NodeValue, ToTraversalPath,
+    },
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+use serde_json::Value;
+use typenum::Unsigned;
+
+use std::{cmp::Ordering, sync::Arc};
+
+// This trait defines methods that a data source should implement
+// It enables retrieval of the membership path for a leaf node, which can be used to reconstruct the Merkle tree state.
+#[async_trait]
+pub trait MerklizedStateDataSource<Types>
+where
+    Types: NodeType,
+{
+    type Error: std::error::Error + std::fmt::Debug + Send + Sync + 'static;
+    async fn get_path<
+        E: Element + Send + DeserializeOwned,
+        I: Index + Send + ToTraversalPath<A> + DeserializeOwned,
+        A: Unsigned,
+        T: NodeValue + Send + CanonicalDeserialize,
+    >(
+        &self,
+        _state_type: &'static str,
+        _tree_height: usize,
+        _header_state_commitment_field: &'static str,
+        _snapshot: Snapshot<Types>,
+        _key: Value,
+    ) -> Result<MerklePath<E, I, T>, Self::Error> {
+        Ok(vec![MerkleNode::Empty])
+    }
+}
+
+// This trait defines methods for updating the storage with the merkle tree state.
+#[async_trait]
+pub trait UpdateStateData: Send + Sync {
+    type Error: std::error::Error + std::fmt::Debug + Send + Sync + 'static;
+    async fn insert_merkle_nodes<
+        Proof: MembershipProof<E, I, T> + Send + Sync + std::fmt::Debug + 'static,
+        E: Element + Send + Sync + Serialize,
+        I: Index + Send + Sync + Serialize,
+        T: NodeValue + Send + Sync,
+    >(
+        &mut self,
+        _name: &'static str,
+        _proof: Proof,
+        _path: Vec<usize>,
+        _block_number: u64,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+// This trait should be implemented by types that represent validated states.
+// It enables services using the module to implement update functionality.
+// Some services may manage multiple Merkle trees within their validated state,
+// requiring updates for all state types with each event.
+// Therefore, the `insert_merkle_nodes` method of `UpdateStateData` should be called
+// for all state types within the validated state.
+#[async_trait]
+pub trait UpdateStateStorage<Types: NodeType> {
```

Comment:
If we end up taking my suggestion above (about making the data source generic over the state type) I think it would be good to parameterize this trait on the data source being updated, e.g. `trait UpdateStateStorage<Types: NodeType, State: MerklizedState, DataSource: MerklizedStateDataSource<Types, State>>`. The reason is that this allows the caller to specify what constraints they require of the data source. E.g. in the sequencer we know we need `DataSource: UpdateStateData<FeeMerkleTree> + UpdateStateData<BlockMerkleTree>`. Then we can _remove_ the `UpdteStateData` constraint from `UpdateDataSource`, which in turn allows us to remove the state trait impls from data sources that don't support them, like file system, while still implementing `UpdateDataSource`. This would prevent us from even compiling code that tried to use the file system storage with a `ValidatedState` type that required support for merklized storage.

Meanwhile, for the mock validated state type, we could implement `UpdateStateStorage` with _no additional constraints_, since it doesn't have any merklized state to store. Then it could be used with any data source.

I think this would be a nice cleanup of the interfaces, lmk what you think.

Reply:
so we would need to remove the state update from the update function currently used for updating the availability data